das/checkpoint_store.go

```diff
@@ -13,35 +13,45 @@ var (
 	checkpointKey = datastore.NewKey("checkpoint")
 )
 
+type checkpoint struct {
+	MinSampled uint64         `json:"min_sampled_height"` // all headers before minSampled were successfully sampled
+	MaxKnown   uint64         `json:"max_known_height"`   // height of the newest known header
+	Failed     map[uint64]int `json:"failed"`             // skipped header's heights with corresponding try count
+	Workers    []workerState  `json:"workers"`
+}
+
 // wrapCheckpointStore wraps the given datastore.Datastore with the `das`
 // prefix. The checkpoint store stores/loads the DASer's checkpoint to/from
-// disk using the checkpointKey. The checkpoint is stored as a uint64
+// disk using the checkpointKey. The checkpoint is stored as an uint64
 // representation of the height of the latest successfully DASed header.
 func wrapCheckpointStore(ds datastore.Datastore) datastore.Datastore {
 	return namespace.Wrap(ds, storePrefix)
 }
 
-// loadCheckpoint loads the DAS checkpoint height from disk and returns it.
+// loadCheckpoint loads the DAS checkpoint from disk and returns it.
 // If there is no known checkpoint, it returns height 0.
-func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (int64, error) {
-	checkpoint, err := ds.Get(ctx, checkpointKey)
+func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (checkpoint, error) {
+	bs, err := ds.Get(ctx, checkpointKey)
 	if err != nil {
-		// if no checkpoint was found, return checkpoint as
-		// 0 since DASer begins sampling on checkpoint+1
+		// if no checkpoint was found, return zero-value checkpoint
 		if err == datastore.ErrNotFound {
 			log.Debug("checkpoint not found, starting sampling at block height 1")
-			return 0, nil
+			return checkpoint{}, nil
 		}
 
-		return 0, err
+		return checkpoint{}, err
 	}
-	return int64(binary.BigEndian.Uint64(checkpoint)), err
+
+	cp := checkpoint{}
+	err = json.Unmarshal(bs, &cp)
+	return cp, err
 }
```

Comment:
What's the point of passing value? The q is not about this place specifically, but about all the places where the checkpoint is copied when passed.

Reply:
You are right, returning and passing by copy mean extra coping. When returning and passing by copy it will be allocated on short-lived stack, on the other hand returning by pointer will require compiler to do heap allocation.

A stack allocation is short-lived and typically very cheap. No Garbage Collection is required for a stack allocation as the end of the function is also the end of the variable’s lifetime. On the other hand, a heap allocation is long-lived and considerably more expensive. When allocating onto the heap, the runtime must find a contiguous piece of memory where the new variable fits. Additionally, it must be garbage-collected when the variable is no longer us

ed. Both operations are orders of magnitudes more expensive than a stack allocation.

To showcase benefits of passing by copy I've recorder 2 test traces on local machine:
[copy_trace.out.zip](https://github.com/celestiaorg/celestia-node/files/9431324/copy_trace.out.zip)
[pointer_trace.out.zip](https://github.com/celestiaorg/celestia-node/files/9431325/pointer_trace.out.zip)
You can run them to investigate allocation details and GC time by 
```
go tool trace pointer_trace.out
```
There is also small benchmark in `coordinator_test.go` you I've run with `-benchmem` flag. Results for pointer vs copy on my local machine:
```
BenchmarkManager/bench_run-copy10         	 7310666	       145.9 ns/op	      18 B/op	       0 allocs/op
BenchmarkManager/bench_run-pointer-10         	 7417952	       182.2 ns/op	      21 B/op	       0 allocs/op
```

Reply:
Thanks for the detailed proof. That's what we were missing in our knowledge, and I am pretty sure there are other places throughout the codebase where we could use value rather than ptrs. Pls report those if you find them or even fix them through PRs 🙏🏻